src/btreemap/node.rs

```diff
@@ -147,34 +152,55 @@ impl<K: Storable + Ord + Clone> Node<K> {
 
     /// Returns true if the node cannot store anymore entries, false otherwise.
     pub fn is_full(&self) -> bool {
-        self.keys.len() >= CAPACITY
+        self.keys_encoded_values.len() >= CAPACITY
     }
 
     /// Swaps the entry at index `idx` with the given entry, returning the old entry.
     pub fn swap_entry<M: Memory>(
         &mut self,
         idx: usize,
-        (mut key, value): Entry<K>,
+        (key, value): Entry<K>,
         memory: &M,
     ) -> Entry<K> {
-        core::mem::swap(&mut self.keys[idx], &mut key);
-        let old_value = self.value(idx, memory).to_vec();
-        self.encoded_values.borrow_mut()[idx] = Value::ByVal(value);
-        (key, old_value)
+        let mut swap_with = (key, RefCell::new(Value::ByVal(value)));
+        core::mem::swap(&mut self.keys_encoded_values[idx], &mut swap_with);
+        (
+            swap_with.0,
+            self.resolve_value(RefCell::into_inner(swap_with.1), memory),
+        )
     }
 
     /// Returns a copy of the entry at the specified index.
     pub fn entry<M: Memory>(&self, idx: usize, memory: &M) -> Entry<K> {
-        (self.keys[idx].clone(), self.value(idx, memory).to_vec())
+        (
+            self.keys_encoded_values[idx].0.clone(),
+            self.value(idx, memory).to_vec(),
+        )
     }
 
     /// Returns a reference to the encoded value at the specified index.
     pub fn value<M: Memory>(&self, idx: usize, memory: &M) -> Ref<[u8]> {
         // Load and cache the value from the underlying memory if needed.
-        {
-            let mut values = self.encoded_values.borrow_mut();
+        let mut value = self.keys_encoded_values[idx].1.borrow_mut();
+        if let Value::ByRef(offset) = *value {
+            // Cache the value internally.
+            *value = Value::ByVal(self.resolve_value(Value::ByRef(offset), memory));
+        }
+        drop(value); // drop borrow because it's reborrowed below
+
+        // Return a reference to the value.
+        Ref::map(self.keys_encoded_values[idx].1.borrow(), |values| {
+            if let Value::ByVal(v) = values {
+                &v[..]
+            } else {
+                unreachable!("value must have been loaded already.");
+            }
+        })
```

Comment:
Did you try to avoid the `unreachable` by restructuring the code and handling things in a single match and it didn't work? Is that why the above code snippet as well? It seems like it (I guess we cannot be iterating over a collection and wanting to change it at the same time). In that case, I suggest the following:

1. Expand the error message a bit to say "have been loaded already a few lines above" or so.
2. Maybe leave a comment that explains that we do this because iterating over the values and wanting to change them at the same time (in case we first want to cache it, then return it) does not work and we are forced to go down the `unreachable` path.

Reply:
Actually, the more I am reading this code, I don't see why we need the `Ref::map`? Can't we just index in the value we want and return that directly?

Reply:
Note that this unreachable() was there already before, I just moved it. I don't think it's possible to remove it if we want to keep the `Ref<[u8]>`, since referencing the value out of the borrow has to happen in the closure passed to `Ref::map`.

`Ref::map` is necessary if we want to avoid cloning the value (the alternative would be returning a `Vec<u8>` and cloning on every invocation). Without `Ref` the borrow would die when the function returns so the returned reference would be invalid.

Reply:
One more alternative in the future would be calling `Ref::leak` which would return an immutable naked reference with the lifetime of the Node. However, this is only available on nightly at the moment.

Reply:
One more alternative would be using `OnceCell` (it was stabilized in Rust 1.80). It would require some refactoring, though. With `OnceCell` we could avoid `unreachable!()` and it would be perhaps a little more expressive than `RefCell`.

Reply:
Thanks for the detailed explanation!

Yeah I know it was there before, I was mainly asking to see if there's anything possible now that you are refactoring some of the code. I believe we can simply keep the unreachable then, no need to get a lot more fancy.

Reply:
Improved the error message and replaced the code inside `Ref::map` to use a `match` with explicit arms. This way we'll get a compile-time error message if we ever add new enum variants to `Value`.